python/aghast/connect/fnalhist.py

```diff
@@ -0,0 +1,173 @@
+            expanded_shape.append(ax.size)
+        elif isinstance(ax, hist.Bin) and not ax._uniform:
+            axes.append(Axis(EdgesBinning(ax._bins[:-1],
+                                        overflow=_ovf_convention()
+                                        ), expression=ax.name, title=ax.label))
+            expanded_shape.append(ax.size)
+        else:
+            raise NotImplementedError("unknown fnalhist Axis type")
+
+    expanded_sumw = numpy.zeros(shape=expanded_shape, dtype=obj._dtype)
+    if obj._sumw2 is not None:
+        expanded_sumw2 = numpy.zeros(shape=expanded_shape, dtype=obj._dtype)
+
+    walk_shape = [len(s) for s in expanded_slices]
+    walk_indices = numpy.unravel_index(numpy.arange(numpy.prod(walk_shape)), walk_shape)
+    extract_slices = list(zip(*(v[k] for v,k in zip(expanded_slices, walk_indices))))
+    for iwalk, tup in enumerate(extract_slices):
+        if tup in obj._sumw:
+            insert_slice = [slice(None) for _ in expanded_shape]
+            for walk_dim, ins_dim in enumerate(expansion_offsets):
+                insert_slice[ins_dim] = walk_indices[walk_dim][iwalk]
+            insert_slice = tuple(insert_slice)
+            expanded_sumw[insert_slice] = obj._sumw[tup]
+            if obj._sumw2 is not None:
+                expanded_sumw2[insert_slice] = obj._sumw2[tup]
+
+
+    if obj._sumw2 is not None:
+        counts = WeightedCounts(
+            sumw=InterpretedInlineBuffer.fromarray(expanded_sumw),
+            sumw2=InterpretedInlineBuffer.fromarray(expanded_sumw2)
+        )
+    else:
+        counts = UnweightedCounts(InterpretedInlineBuffer.fromarray(expanded_sumw))
+
+    hout = Histogram(axis=axes, title=obj.label, counts=counts)
+    return hout
+
+
+def tofnalhist(obj):
+    if not isinstance(obj, Histogram):
+        raise TypeError("cannot convert {0} to a fnal_column_analysis_tools histogram".format(type(obj).__name__))
+
+    axes = []
+    sparse_binning = {}
+    dense_map = {}
+    dense_shape = []
+    for i, ax in enumerate(obj.axis):
+        if isinstance(ax.binning, CategoryBinning):
+            sparse_binning[i] = ax.binning.categories
+            if ax.binning.loc_overflow == BinLocation.below1:
+                sparse_binning[i].insert(0, "")
+            elif ax.binning.loc_overflow == BinLocation.above1:
+                sparse_binning[i].append("")
+            elif ax.binning.loc_overflow != BinLocation.nonexistent:
+                warnings.warn("why would you have the sole overflow bin not adjacent to normal range?!", RuntimeWarning)
+            new_ax = hist.Cat(ax.expression, ax.title)
+            new_ax._categories = sparse_binning[i]
+            axes.append(new_ax)
+        elif isinstance(ax.binning, RegularBinning):
+            if not ax.binning.interval.low_inclusive:
+                warnings.warn("fnal_column_analysis_tools hist types do not understand non-low-inclusive binning", RuntimeWarning)
```

Comment:
Just ignore this case (consistency with `toroot`/`fromroot`). Aghast carries the metadata about how it was filled, and that can be useful for determining why a translation to another format gives a different result than natively filling the histogram in that format.

There are three choices for what we could do here.
    * **ignore it:** It's a single floating point value. The only way to get a difference in histograms from this is if the data being histogrammed were not continuously distributed but peaked at the edge value. You can get differences on this scale from other things we don't control, like non-associativity of floating point numbers.
    * **forbid it:** If we were to be strict about it, we would have to full stop here. There is no way to port a right-edge histogram to a left-edge histogram if we care about binary compatibility, and not just continuous approximation.
    * **warnings:** There _will_ be many cases in which the user never gets to see the stderr stream. Histograms will be combined remotely on Spark. On the other hand, when we do see them, it's often in a context where it's undesirable. If the user is directly using a library and knowing what's supposed to happen, warnings can be helpful; but if it's something buried a layer of abstraction deep, the warnings only cause confusion.

For instance, you're working directly with Numpy and you divide by zero, see the warning, and that's great. But often some framework is using Numpy, perhaps without your knowledge, and it starts spewing division by zero warnings? How does that help? I think with Aghast, we'll often be in the latter case.